lib/rucio/core/replica.py

```diff
@@ -1830,17 +1830,22 @@ def _expected_pfns(lfns, rse_settings, scheme, operation='write', domain='wan',
                 # Check that the pfns match to the expected pfns
                 lfns = [{'scope': i['scope'].external, 'name': i['name']} for i in files if i['pfn'].startswith(scheme)]
                 pfns[scheme] = clean_pfns(pfns[scheme])
+                expected_pfns = set()
 
                 for protocol_attr in rsemgr.get_protocols_ordered(rse_settings=rse_settings, operation='write', scheme=scheme, domain='wan'):
-                    pfns[scheme] = list(set(pfns[scheme]) - set(_expected_pfns(lfns, rse_settings, scheme, operation='write', domain='wan', protocol_attr=protocol_attr)))
+                    expected_pfn = set(_expected_pfns(lfns, rse_settings, scheme, operation='write', domain='wan', protocol_attr=protocol_attr))
+                    expected_pfns = expected_pfns.union(expected_pfn)
+                    pfns[scheme] = list(set(pfns[scheme]) - expected_pfn)
```

Comment:
Couldn't this be placed after the for loop? i.e. instead of doing `set(pfns[scheme]) - expected_pfn` for every `expected_pfn`, we just do it all at once after the loop by doing `set(pfns[scheme]) - expected_pfns`